NotificationHubs/src/com/windowsazure/messaging/NotificationHubJob.java

```diff
@@ -7,178 +11,303 @@
-		buf.append(XML_HEADER);
-		if(this.jobType!=null) buf.append("<Type>" + this.jobType.name()+ "</Type>");
-		if(this.outputContainerUri!=null) buf.append("<OutputContainerUri>" + this.outputContainerUri + "</OutputContainerUri>");
-		if(this.importFileUri!=null) buf.append("<ImportFileUri>" + this.importFileUri + "</ImportFileUri>");
-		buf.append(XML_FOOTER);
-		return buf.toString();
-	}	
-	
-	private static void setupCollectionParser(Digester digester){	
-		digester.addObjectCreate("*/feed", LinkedList.class);
-		setupSingleEntryParser(digester);
-		digester.addSetNext("*/entry", "add", NotificationHubJob.class.getName());
-	}
-	
-	private static void setupSingleEntryParser(Digester digester){
-		digester.addObjectCreate("*/entry", NotificationHubJob.class);
-		digester.addCallMethod("*/JobId", "setJobId",1);
-		digester.addCallParam("*/JobId", 0);
-		digester.addCallMethod("*/Progress", "setProgressFromString",1);
-		digester.addCallParam("*/Progress", 0);
-		digester.addCallMethod("*/Type", "setJobTypeFromString",1);
-		digester.addCallParam("*/Type", 0);
-		digester.addCallMethod("*/Status", "setJobStatusFromString",1);
-		digester.addCallParam("*/Status", 0);
-		digester.addCallMethod("*/OutputContainerUri", "setOutputContainerUri",1);
-		digester.addCallParam("*/OutputContainerUri", 0);
-		digester.addCallMethod("*/ImportFileUri", "setImportFileUri",1);
-		digester.addCallParam("*/ImportFileUri", 0);
-		digester.addCallMethod("*/Failure", "setFailure",1);
-		digester.addCallParam("*/Failure", 0);
-		digester.addCallMethod("*/CreatedAt", "setCreatedAtFromString",1);
-		digester.addCallParam("*/CreatedAt", 0);
-		digester.addCallMethod("*/UpdatedAt", "setUpdatedAtFromString",1);
-		digester.addCallParam("*/UpdatedAt", 0);		
-		digester.addObjectCreate("*/OutputProperties", HashMap.class);
-		digester.addCallMethod("*/d3p1:KeyValueOfstringstring", "put",2);
-		digester.addCallParam("*/d3p1:Key", 0);
-		digester.addCallParam("*/d3p1:Value", 1);
-		digester.addSetNext("*/OutputProperties", "setOutputProperties", Map.class.getName());
-	}
+    private static final String XML_HEADER = "<?xml version=\"1.0\" encoding=\"utf-8\"?><entry xmlns=\"http://www.w3.org/2005/Atom\"><content type=\"application/atom+xml;type=entry;charset=utf-8\"><NotificationHubJob xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://schemas.microsoft.com/netservices/2010/10/servicebus/connect\">";
+    private static final String XML_FOOTER = "</NotificationHubJob></content></entry>";
+
+    private String jobId;
+    private double progress;
+    private NotificationHubJobType jobType;
+    private NotificationHubJobStatus jobStatus;
+    private String outputContainerUri;
+    private String importFileUri;
+    private String failure;
+    private Map<String, String> outputProperties;
+    private Date createdAt;
+    private Date updatedAt;
+
+    private static final AtomicReference<Digester> singleEntryParser = new AtomicReference<>();
```

Comment:
Same question on the difference between ThreadLocal and AtomicReference.

Reply:
Let's go with Atomic since it makes more sense here instead of being a thread local which didn't really make much sense in terms of being a singleton